File: code/ex127.java
// BFS

class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> dict = new HashSet<>(wordList);
        if (!dict.contains(endWord))    return 0;
        
        int layer = 0;
        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);    layer++;
        
        while (!queue.isEmpty()) {
            int queue_size = queue.size();

            for (int q = 0; q < queue_size; ++q) {    // Loop through a layer
                String out = queue.poll();
            
                for (int i = 0; i < out.length(); i++) {    // Loop through each digits
                    
                    for (char c = 'a'; c <= 'z'; c++) {    // Loop each char 'a'-'z' to replace that digit
                        String newstring = out.substring(0, i) + c + out.substring(i + 1);
                        if (!dict.contains(newstring))    continue;
                        
                        if (newstring.equals(endWord))    return layer + 1;
                        else {
                            queue.offer(newstring);
                            dict.remove(newstring);
                        }
                    }
                }
            }
            layer++;
        }

        return 0;
    }
}

File: code/ex127.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> dict(wordList.begin(), wordList.end());
        if (dict.count(endWord) == 0)    return 0;
        
        queue<string> que;
        que.push(beginWord);
        
        int step = 0;
        while (!que.empty()) {
            int size = que.size();
            while (size--) {
                string out = que.front();    que.pop();
                if (out == endWord)    return step + 1;
                for (int i = 0; i < out.length(); i++) {
                    char c = out[i];
                    for (char nc = 'a'; nc <= 'z'; nc++) {
                        out[i] = nc;
                        if (dict.count(out) > 0) {
                            dict.erase(out);
                            que.push(out);
                        }
                    }
                    out[i] = c;
                }
            }
            step++;
        }
        
        return 0;
    }
};